How to properly handle "unhandled" exceptions

### Describe the bug

I have a test that asserts that an import throws, because the imported module is not found.

even though the actual error is caught by `expect().rejects.toThrowError()` it still shows up as unhandled.

```
⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯ Unhandled Errors ⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯

Vitest caught 1 unhandled error during the test run.
This might cause false positive tests. Resolve unhandled errors to make sure your tests are not affected.

⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯ Unhandled Rejection ⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯

Error: Cannot find module 'does-not-exist'
Require stack:
- /private/var/folders/y3/2b5xsndd3zvd32j6n_ypjg5r0000gn/T/zip-it-test-bundler-nft-31679-4ePKnlfzcJsT/node-module-missing/function.js
- /var/folders/y3/2b5xsndd3zvd32j6n_ypjg5r0000gn/T/zip-it-test-bundler-nft-31679-4ePKnlfzcJsT/function.js
 ❯ Function.Module._resolveFilename node:internal/modules/cjs/loader:995:15
```

I tried marking all the folders as external, but that does nothing.

There is also this message:
```
✘ [ERROR] Could not resolve "does-not-exist"

    tests/fixtures/node-module-missing/function.js:1:25:
      1 │ module.exports = require('does-not-exist')
        ╵                          ~~~~~~~~~~~~~~~~

  You can mark the path "does-not-exist" as external to exclude it from the bundle, which will
  remove this error. You can also surround this "require" call with a try/catch block to handle this
  failure at run-time instead of bundle-time.

```
which does not fail any test, but I also cannot get rid of, even if I set the importing file as external.

### Reproduction

Minimal repro:
https://github.com/danez-labs/vitest-import-unhandled

- `npm i`
- `npm test`

Or the original problem:

https://github.com/netlify/zip-it-and-ship-it/pull/1286

- `npm i`
- `npm exec vitest -- run -t 'Throws on missing dependencies '`

### System Info

```shell
System:
    OS: macOS 12.6.1
    CPU: (10) arm64 Apple M1 Pro
    Memory: 2.90 GB / 32.00 GB
    Shell: 5.8.1 - /bin/zsh
  Binaries:
    Node: 18.12.1 - ~/.asdf/installs/nodejs/lts/bin/node
    Yarn: 1.22.19 - /opt/homebrew/bin/yarn
    npm: 8.19.2 - ~/.asdf/plugins/nodejs/shims/npm
  Browsers:
    Safari: 16.1
```


### Used Package Manager

npm

### Validations

- [X] Follow our [Code of Conduct](https://github.com/vitest-dev/vitest/blob/main/CODE_OF_CONDUCT.md)
- [X] Read the [Contributing Guidelines](https://github.com/vitest-dev/vitest/blob/main/CONTRIBUTING.md).
- [X] Read the [docs](https://vitest.dev/guide/).
- [X] Check that there isn't [already an issue](https://github.com/vitest-dev/vitest/issues) that reports the same bug to avoid creating a duplicate.
- [X] Check that this is a concrete bug. For Q&A open a [GitHub Discussion](https://github.com/vitest-dev/vitest/discussions) or join our [Discord Chat Server](https://chat.vitest.dev).
- [X] The provided reproduction is a [minimal reproducible example](https://stackoverflow.com/help/minimal-reproducible-example) of the bug.

--- a/packages/vite-node/src/client.ts
+++ b/packages/vite-node/src/client.ts
@@ -180,11 +180,12 @@ export class ViteNodeRunner {
     const promise = this.directRequest(id, fsPath, callstack)
     Object.assign(mod, { promise, evaluated: false })
 
-    promise.finally(() => {
+    try {
+      return await promise
+    }
+    finally {
       mod.evaluated = true
-    })
-
-    return await promise
+    }
   }
 
   /** @internal */
@@ -272,7 +273,7 @@ export class ViteNodeRunner {
       throw new Error(`[vite-node] Failed to load ${id}`)
 
     const file = cleanUrl(resolvedId || fsPath)
-    // console.log('file', file)
+
     // disambiguate the `<UNIT>:/` on windows: see nodejs/node#31710
     const url = pathToFileURL(file).href
     const meta = { url }

--- a/test/core/test/imports.test.ts
+++ b/test/core/test/imports.test.ts
@@ -53,3 +53,9 @@ test('dynamic import has null prototype', async () => {
 
   expect(Object.getPrototypeOf(stringTimeoutMod)).toBe(null)
 })
+
+test('dynamic imports throws an error', async () => {
+  const path = './some-unknown-path'
+  const imported = import(path)
+  await expect(imported).rejects.toThrowError(/Failed to load/)
+})

--- a/test/core/test/imports.test.ts
+++ b/test/core/test/imports.test.ts
@@ -54,8 +54,10 @@ test('dynamic import has null prototype', async () => {
   expect(Object.getPrototypeOf(stringTimeoutMod)).toBe(null)
 })
 
-test('dynamic imports throws an error', async () => {
+test('dynamic import throws an error', async () => {
   const path = './some-unknown-path'
   const imported = import(path)
   await expect(imported).rejects.toThrowError(/Failed to load/)
+  // @ts-expect-error path does not exist
+  await expect(() => import('./some-unknown-path')).rejects.toThrowError(/Failed to load/)
 })

